### PR TITLE
Use oci-go-sdk in place of custom client

### PR DESCRIPTION
**Pull request by harveylowndes**
Removed custom client replacing with with oci-go-sdk.

### CODE REVIEW

File: builder/oracle/oci/config.go [hunk 1]
```diff
@@ -61,140 +62,96 @@ func NewConfig(raws ...interface{}) (*Config, error) {
...
-		accessCfg.Tenancy = c.TenancyID
-	}
-
-	if c.Region != "" {
-		accessCfg.Region = c.Region
-	}
-
-	// Default if the template nor the API config contains a region.
-	if accessCfg.Region == "" {
-		accessCfg.Region = "us-phoenix-1"
```
Comment (prydie) [inline builder/oracle/oci/config.go]: Maintain defaulting by constructing fileProvider first and calling `fileProvider.Region()` `if c.Region == ""` and defaulting if err / `value == ""`

File: builder/oracle/oci/config.go [hunk 2]
```diff
@@ -61,140 +62,96 @@ func NewConfig(raws ...interface{}) (*Config, error) {
...
 	}
 
-	if c.KeyFile != "" {
-		accessCfg.KeyFile = c.KeyFile
-		accessCfg.Key, err = client.LoadPrivateKey(accessCfg)
-		if err != nil {
-			return nil, fmt.Errorf("Failed to load private key %s : %s", accessCfg.KeyFile, err)
-		}
+	// Load API access configuration from SDK
+	accessCfg, err := conf.ComposingConfigurationProvider(providers)
```
Comment (prydie) [inline builder/oracle/oci/config.go]: `s/accessCfg/configProvider/`

File: builder/oracle/oci/config.go [hunk 3]
```diff
@@ -61,140 +62,96 @@ func NewConfig(raws ...interface{}) (*Config, error) {
...
-	if accessCfg.Fingerprint == "" {
+	fingerprint, _ := accessCfg.KeyFingerprint()
+	if fingerprint == "" {
 		errs = packer.MultiErrorAppend(
 			errs, errors.New("'fingerprint' must be specified"))
 	}
 
-	if accessCfg.Key == nil {
+	//TODO (HarveyLowndes) when does this condition occur?
+	privateKey := accessCfg.PrivateRSAKey
```
Comment (prydie) [inline builder/oracle/oci/config.go]:
```go
if _, err := accessCfg.PrivateRSAKey(); err != nil {
```

File: builder/oracle/oci/config.go [hunk 4]
```diff
@@ -61,140 +62,96 @@ func NewConfig(raws ...interface{}) (*Config, error) {
...
 		errs = packer.MultiErrorAppend(
-			errs, errors.New("'key_file' must be specified"))
+			errs, errors.New("'PrivateRSAKey' must be specified"))
 	}
 
 	c.AccessCfg = accessCfg
 
-	// Required non AccessCfg configuration options
-
-	if c.AvailabilityDomain == "" {
```
Comment (prydie) [inline builder/oracle/oci/config.go]: Re-add all of these

File: builder/oracle/oci/config_test.go [hunk 5]
```diff
@@ -1,50 +1,61 @@
...
+	"encoding/pem"
 	"io/ioutil"
 	"os"
 	"reflect"
 	"strings"
 	"testing"
 
-	client "github.com/hashicorp/packer/builder/oracle/oci/client"
+	"github.com/go-ini/ini"
+	conf "github.com/oracle/oci-go-sdk/common"
```
Comment (prydie) [inline builder/oracle/oci/config_test.go]: `s/conf/ocicommon/`

File: builder/oracle/oci/config_test.go [hunk 6]
```diff
@@ -1,50 +1,61 @@
...
 		"use_private_ip": false,
 	}
 }
 
-func getField(c *client.Config, field string) string {
+func getField(c *conf.ConfigurationProvider, field string) string {
 	r := reflect.ValueOf(c)
 	f := reflect.Indirect(r).FieldByName(field)
-	return string(f.String())
+	return string(f.String()) //TODO(harveylowndes) Wrap with string?
```
Comment (prydie) [inline builder/oracle/oci/config_test.go]: Undo this TODO

File: builder/oracle/oci/driver_oci.go [hunk 7]
```diff
@@ -1,120 +1,204 @@
...
-	client "github.com/hashicorp/packer/builder/oracle/oci/client"
+	core "github.com/oracle/oci-go-sdk/core"
 )
 
 // driverOCI implements the Driver interface and communicates with Oracle
 // OCI.
 type driverOCI struct {
-	client *client.Client
-	cfg    *Config
+	client    core.ComputeClient
```
Comment (prydie) [inline builder/oracle/oci/driver_oci.go]: `s/client/computeClient/`

File: builder/oracle/oci/driver_oci.go [hunk 8]
```diff
@@ -1,120 +1,204 @@
...
+	vcnclient core.VirtualNetworkClient
```
Comment (prydie) [inline builder/oracle/oci/driver_oci.go]: `s/vcnclient/vcnClient/g`

File: builder/oracle/oci/driver_oci.go [hunk 9]
```diff
@@ -1,120 +1,204 @@
...
-	params := &client.LaunchInstanceParams{
-		AvailabilityDomain: d.cfg.AvailabilityDomain,
-		CompartmentID:      d.cfg.CompartmentID,
-		ImageID:            d.cfg.BaseImageID,
-		Shape:              d.cfg.Shape,
-		SubnetID:           d.cfg.SubnetID,
+	// This should be passed through the parameters however unsure where if to put this in.
+	ctx := context.TODO()
+
+	instance, err := d.client.LaunchInstance(ctx, core.LaunchInstanceRequest{LaunchInstanceDetails: core.LaunchInstanceDetails{
```
Comment (prydie) [inline builder/oracle/oci/driver_oci.go]: `s/ctx/context.TODO()/`

File: builder/oracle/oci/driver_oci.go [hunk 10]
```diff
@@ -1,120 +1,204 @@
...
-		CompartmentID: d.cfg.CompartmentID,
-		InstanceID:    id,
-		DisplayName:   d.cfg.ImageName,
-	}
-	image, err := d.client.Compute.Images.Create(params)
+func (d *driverOCI) CreateImage(id string) (core.Image, error) {
+	// This should be passed through the parameters however unsure where if to put this in.
+	ctx := context.TODO()
+
+	res, err := d.client.CreateImage(ctx, core.CreateImageRequest{CreateImageDetails: core.CreateImageDetails{
```
Comment (prydie) [inline builder/oracle/oci/driver_oci.go]: `s/ctx/context.TODO()/`

File: builder/oracle/oci/driver_oci.go [hunk 11]
```diff
@@ -1,120 +1,204 @@
...
+	return res.Image, nil
 }
 
 // DeleteImage deletes a custom image.
 func (d *driverOCI) DeleteImage(id string) error {
-	return d.client.Compute.Images.Delete(&client.DeleteImageParams{ID: id})
+	// This should be passed through the parameters however unsure where if to put this in.
+	ctx := context.TODO()
+
+	_, err := d.client.DeleteImage(ctx, core.DeleteImageRequest{ImageId: &id})
```
Comment (prydie) [inline builder/oracle/oci/driver_oci.go]: `s/ctx/context.TODO()/`

File: builder/oracle/oci/driver_oci.go [hunk 12]
```diff
@@ -1,120 +1,204 @@
...
-	vnics, err := d.client.Compute.VNICAttachments.List(
-		&client.ListVnicAttachmentsParams{
-			InstanceID:    id,
-			CompartmentID: d.cfg.CompartmentID,
-		},
-	)
+	// This should be passed through the parameters however unsure where if to put this in.
+	ctx := context.TODO()
+
+	vnics, err := d.client.ListVnicAttachments(ctx, core.ListVnicAttachmentsRequest{
```
Comment (prydie) [inline builder/oracle/oci/driver_oci.go]: `s/ctx/context.TODO()/`

File: builder/oracle/oci/driver_oci.go [hunk 13]
```diff
@@ -1,120 +1,204 @@
...
 }
 
 // TerminateInstance terminates a compute instance.
 func (d *driverOCI) TerminateInstance(id string) error {
-	params := &client.TerminateInstanceParams{ID: id}
-	return d.client.Compute.Instances.Terminate(params)
+	// This should be passed through the parameters however unsure where if to put this in.
+	ctx := context.TODO()
+
+	_, err := d.client.TerminateInstance(ctx, core.TerminateInstanceRequest{
```
Comment (prydie) [inline builder/oracle/oci/driver_oci.go]: `s/ctx/context.TODO()/`

File: builder/oracle/oci/driver_oci.go [hunk 14]
```diff
@@ -1,120 +1,204 @@
...
+			instance, err := d.client.GetInstance(ctx, core.GetInstanceRequest{InstanceId: &id})
+			if err != nil {
+				return "", err
+			}
+			return string(instance.LifecycleState), nil
+		},
 		id,
 		waitStates,
 		terminalState,
+		20,    //Temporary
```
Comment (prydie) [inline builder/oracle/oci/driver_oci.go]: Reset to match previous behavior

File: builder/oracle/oci/builder.go [hunk 15]
```diff
@@ -78,10 +79,16 @@ func (b *Builder) Run(ui packer.Ui, hook packer.Hook, cache packer.Cache) (packe
 		return nil, rawErr.(error)
 	}
 
+	region, err := b.config.AccessCfg.Region()
+
```
Comment (prydie) [inline builder/oracle/oci/builder.go]: ~~`\n`~~

File: builder/oracle/oci/driver_oci.go [hunk 16]
```diff
@@ -1,120 +1,204 @@
...
+			image, err := d.client.GetImage(ctx, core.GetImageRequest{ImageId: &id})
+			if err != nil {
+				return "", err
+			}
+			return string(image.LifecycleState), nil //TODO(HarveyLowndes) should these be wrapped?
+		},
 		id,
 		[]string{"PROVISIONING"},
 		"AVAILABLE",
+		20,    //Temporary
```
Comment (prydie) [inline builder/oracle/oci/driver_oci.go]: Return to previous

File: builder/oracle/oci/driver_oci.go [hunk 17]
```diff
@@ -1,120 +1,204 @@
...
+	// This should be passed through the parameters however unsure where if to put this in.
+	ctx := context.TODO()
+
+	return waitForResourceToReachState(
+		func(string) (string, error) {
+			image, err := d.client.GetImage(ctx, core.GetImageRequest{ImageId: &id})
+			if err != nil {
+				return "", err
+			}
+			return string(image.LifecycleState), nil //TODO(HarveyLowndes) should these be wrapped?
```
Comment (prydie) [inline builder/oracle/oci/driver_oci.go]: remove comment

File: builder/oracle/oci/driver_oci.go [hunk 18]
```diff
@@ -1,120 +1,204 @@
...
 // "AVAILABLE" state.
 func (d *driverOCI) WaitForImageCreation(id string) error {
-	return client.NewWaiter().WaitForResourceToReachState(
-		d.client.Compute.Images,
+	// This should be passed through the parameters however unsure where if to put this in.
+	ctx := context.TODO()
+
+	return waitForResourceToReachState(
+		func(string) (string, error) {
+			image, err := d.client.GetImage(ctx, core.GetImageRequest{ImageId: &id})
```
Comment (prydie) [inline builder/oracle/oci/driver_oci.go]: `s/ctx/context.TODO()/`

File: builder/oracle/oci/driver_oci.go [hunk 19]
```diff
@@ -1,120 +1,204 @@
...
 // state.
 func (d *driverOCI) WaitForInstanceState(id string, waitStates []string, terminalState string) error {
-	return client.NewWaiter().WaitForResourceToReachState(
-		d.client.Compute.Instances,
+	// This should be passed through the parameters however unsure where if to put this in.
+	ctx := context.TODO()
+
+	return waitForResourceToReachState(
+		func(string) (string, error) {
+			instance, err := d.client.GetInstance(ctx, core.GetInstanceRequest{InstanceId: &id})
```
Comment (prydie) [inline builder/oracle/oci/driver_oci.go]: `s/ctx/context.TODO()/`

File: builder/oracle/oci/artifact.go [hunk 20]
```diff
@@ -26,13 +26,13 @@ func (a *Artifact) Files() []string {
...
 // Id returns the OCID of the associated Image.
 func (a *Artifact) Id() string {
-	return a.Image.ID
+	return *a.Image.Id
 }
 
 func (a *Artifact) String() string {
 	return fmt.Sprintf(
 		"An image was created: '%v' (OCID: %v) in region '%v'",
-		a.Image.DisplayName, a.Image.ID, a.Region,
```
Comment (harveylowndes) [inline builder/oracle/oci/artifact.go]: `a.Image.DisplayName, a.Image.ID, a.Region` will return memory location. Needs derefrencing.

File: builder/oracle/oci/artifact.go [hunk 21]
```diff
@@ -26,13 +26,13 @@ func (a *Artifact) Files() []string {
...
+		a.Image.DisplayName, a.Image.Id, a.Region,
```
Comment (harveylowndes) [inline builder/oracle/oci/artifact.go]: `a.Image.DisplayName, a.Image.Id, a.Region` will return memory location. Needs derefrencing.

Comment (harveylowndes) [inline builder/oracle/oci/artifact.go]: `a.Image.DisplayName, a.Image.Id, a.Region` will return memory location. Needs dereferencing.

File: builder/oracle/oci/config_test.go [hunk 22]
```diff
@@ -169,28 +186,103 @@ func TestConfig(t *testing.T) {
 
 	// Test that AccessCfgFile properties are overridden by their
 	// corresponding template keys.
-	accessOverrides := map[string]string{
-		"user_ocid":    "User",
-		"tenancy_ocid": "Tenancy",
-		"region":       "Region",
-		"fingerprint":  "Fingerprint",
+	// TODO: needs rewriting as individual tests that call the methods on the
```
Comment (harveylowndes) [inline builder/oracle/oci/config_test.go]: Create individual tests for overrides.

File: builder/oracle/oci/driver_oci.go [hunk 23]
```diff
@@ -1,123 +1,189 @@
...
+	for i := 0; maxRetries == 0 || i < maxRetries; i++ {
+
+		state, err := GetResourceState(id)
+
+		if err != nil {
+			return err
+		}
+
+		if stringSliceContains(waitStates, state) {
+			time.Sleep(time.Duration(waitDuration) * time.Millisecond)
```
Comment (prydie) [inline builder/oracle/oci/driver_oci.go]: Becomes `time.Sleep(waitDuration)` with above change

File: builder/oracle/oci/config.go [hunk 24]
```diff
@@ -67,113 +67,110 @@ func NewConfig(raws ...interface{}) (*Config, error) {
...
+		return nil, err
 	}
 
 	var errs *packer.MultiError
 	if es := c.Comm.Prepare(&c.ctx); len(es) > 0 {
 		errs = packer.MultiErrorAppend(errs, es...)
 	}
 
-	// Required AccessCfg configuration options
+	if c.ImageName == "" {
```
Comment (prydie) [inline builder/oracle/oci/config.go]: Move this back to where it was previously to reduce the scope of changes

File: builder/oracle/oci/config.go [hunk 25]
```diff
@@ -236,6 +223,7 @@ func NewConfig(raws ...interface{}) (*Config, error) {
 // config file location ($HOME/.oci/config).
 func getDefaultOCISettingsPath() (string, error) {
 	home, err := homedir.Dir()
+
```
Comment (prydie) [inline builder/oracle/oci/config.go]: ~~`\n`~~

File: builder/oracle/oci/config_test.go [hunk 26]
```diff
@@ -81,14 +84,15 @@ func TestConfig(t *testing.T) {
...
 		_, errs := NewConfig(raw)
 
-		s := errs.Error()
 		expectedErrors := []string{
 			"'user_ocid'", "'tenancy_ocid'", "'fingerprint'",
-			"'key_file'",
 		}
+
+		s := errs.Error()
+
```
Comment (prydie) [inline builder/oracle/oci/config_test.go]: ~~`\n`~~

File: builder/oracle/oci/config_test.go [hunk 27]
```diff
@@ -116,9 +120,15 @@ func TestConfig(t *testing.T) {
 			t.Fatalf("err: %+v", errs)
 		}
 
-		expected := "ocid1.tenancy.oc1..aaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaaa"
-		if c.AccessCfg.Tenancy != expected {
-			t.Errorf("Expected tenancy: %s, got %s.", expected, c.AccessCfg.Tenancy)
+		tenancy, err := c.ConfigProvider.TenancyOCID()
+
```
Comment (prydie) [inline builder/oracle/oci/config_test.go]: ~~`\n`~~

File: builder/oracle/oci/config_test.go [hunk 28]
```diff
@@ -130,9 +140,15 @@ func TestConfig(t *testing.T) {
 			t.Fatalf("err: %+v", errs)
 		}
 
+		region, err := c.ConfigProvider.Region()
+
```
Comment (prydie) [inline builder/oracle/oci/config_test.go]: ~~`\n`~~

File: builder/oracle/oci/config_test.go [hunk 29]
```diff
@@ -159,38 +175,150 @@ func TestConfig(t *testing.T) {
...
+		accessVal, _ := c.ConfigProvider.Region()
+		if accessVal != expected {
+			t.Errorf("Expected ConfigProvider.Region: %s, got %s", expected, accessVal)
+		}
+	})
+
+	t.Run("ConfigProvider.KeyFingerprintOveridden", func(t *testing.T) {
+		expected := "override"
+
```
Comment (prydie) [inline builder/oracle/oci/config_test.go]: ~~`\n`~~

File: builder/oracle/oci/config_test.go [hunk 30]
```diff
@@ -159,38 +175,150 @@ func TestConfig(t *testing.T) {
...
+		accessVal, _ := c.ConfigProvider.TenancyOCID()
+		if accessVal != expected {
+			t.Errorf("Expected ConfigProvider.TenancyOCID: %s, got %s", expected, accessVal)
+		}
+	})
+
+	t.Run("ConfigProvider.RegionOveridden", func(t *testing.T) {
+		expected := "override"
+
```
Comment (prydie) [inline builder/oracle/oci/config_test.go]: ~~`\n`~~

File: builder/oracle/oci/config_test.go [hunk 31]
```diff
@@ -159,38 +175,150 @@ func TestConfig(t *testing.T) {
...
+		accessVal, _ := c.ConfigProvider.UserOCID()
+		if accessVal != expected {
+			t.Errorf("Expected ConfigProvider.UserOCID: %s, got %s", expected, accessVal)
+		}
+	})
+
+	t.Run("ConfigProvider.TenancyOCIDOveridden", func(t *testing.T) {
+		expected := "override"
+
```
Comment (prydie) [inline builder/oracle/oci/config_test.go]: ~~`\n`~~

File: builder/oracle/oci/builder.go [hunk 32]
```diff
@@ -37,6 +37,7 @@ func (b *Builder) Prepare(rawConfig ...interface{}) ([]string, error) {
 }
 
 func (b *Builder) Run(ui packer.Ui, hook packer.Hook, cache packer.Cache) (packer.Artifact, error) {
+
```
Comment (prydie) [inline builder/oracle/oci/builder.go]: ~~`\n`~~

File: builder/oracle/oci/driver_oci.go [hunk 33]
```diff
@@ -1,123 +1,189 @@
...
 
-	return image, nil
+	return res.Image, nil
 }
 
 // DeleteImage deletes a custom image.
 func (d *driverOCI) DeleteImage(id string) error {
-	return d.client.Compute.Images.Delete(&client.DeleteImageParams{ID: id})
+	_, err := d.computeClient.DeleteImage(context.TODO(), core.DeleteImageRequest{ImageId: &id})
+
```
Comment (prydie) [inline builder/oracle/oci/driver_oci.go]: ~~`\n`~~

File: builder/oracle/oci/driver_oci.go [hunk 34]
```diff
@@ -1,123 +1,189 @@
...
 	}
 
-	if len(vnics) < 1 {
+	if len(vnics.Items) < 1 {
 		return "", errors.New("instance has zero VNICs")
 	}
 
-	vnic, err := d.client.Compute.VNICs.Get(&client.GetVNICParams{ID: vnics[0].VNICID})
+	vnic, err := d.vcnClient.GetVnic(context.TODO(), core.GetVnicRequest{VnicId: vnics.Items[0].VnicId})
+
```
Comment (prydie) [inline builder/oracle/oci/driver_oci.go]: ~~`\n`~~

File: builder/oracle/oci/driver_oci.go [hunk 35]
```diff
@@ -1,123 +1,189 @@
...
 	if err != nil {
 		return "", fmt.Errorf("Error getting VNIC details: %s", err)
 	}
 
 	if d.cfg.UsePrivateIP {
-		return vnic.PrivateIP, nil
+		return *vnic.PrivateIp, nil
 	}
-	return vnic.PublicIP, nil
+
```
Comment (prydie) [inline builder/oracle/oci/driver_oci.go]: ~~`\n`~~

File: builder/oracle/oci/driver_oci.go [hunk 36]
```diff
@@ -1,123 +1,189 @@
...
 
 // TerminateInstance terminates a compute instance.
 func (d *driverOCI) TerminateInstance(id string) error {
-	params := &client.TerminateInstanceParams{ID: id}
-	return d.client.Compute.Instances.Terminate(params)
+
+	_, err := d.computeClient.TerminateInstance(context.TODO(), core.TerminateInstanceRequest{
+		InstanceId: &id,
+	})
+
```
Comment (prydie) [inline builder/oracle/oci/driver_oci.go]: ~~`\n`~~

File: builder/oracle/oci/driver_oci.go [hunk 37]
```diff
@@ -1,123 +1,189 @@
...
-	return vnic.PublicIP, nil
+
+	return *vnic.PublicIp, nil
 }
 
 // TerminateInstance terminates a compute instance.
 func (d *driverOCI) TerminateInstance(id string) error {
-	params := &client.TerminateInstanceParams{ID: id}
-	return d.client.Compute.Instances.Terminate(params)
+
```
Comment (prydie) [inline builder/oracle/oci/driver_oci.go]: ~~`\n`~~

File: builder/oracle/oci/driver_oci.go [hunk 38]
```diff
@@ -1,123 +1,189 @@
...
+			if err != nil {
+				return "", err
+			}
+			return string(image.LifecycleState), nil
+		},
 		id,
 		[]string{"PROVISIONING"},
 		"AVAILABLE",
+		0,    //Unlimited Retries
+		5000, //5 second wait between retries
```
Comment (prydie) [inline builder/oracle/oci/driver_oci.go]: `s/5000/5 * time.Second/`

File: builder/oracle/oci/driver_oci.go [hunk 39]
```diff
@@ -1,123 +1,189 @@
...
+			if err != nil {
+				return "", err
+			}
+			return string(instance.LifecycleState), nil
+		},
 		id,
 		waitStates,
 		terminalState,
+		0,    //Unlimited Retries
+		5000, //5 second wait between retries
```
Comment (prydie) [inline builder/oracle/oci/driver_oci.go]: `s/5000/5 * time.Second/`

File: builder/oracle/oci/config_test.go [hunk 40]
```diff
@@ -159,38 +175,150 @@ func TestConfig(t *testing.T) {
...
+			t.Fatalf("err: %+v", errs)
+		}
+
+		accessVal, _ := c.ConfigProvider.TenancyOCID()
+		if accessVal != expected {
+			t.Errorf("Expected ConfigProvider.TenancyOCID: %s, got %s", expected, accessVal)
+		}
+	})
+
+	t.Run("ConfigProvider.RegionOveridden", func(t *testing.T) {
```
Comment (prydie) [inline builder/oracle/oci/config_test.go]: `s/ConfigProvider.RegionOveridden/region_overridden/`

File: builder/oracle/oci/config_test.go [hunk 41]
```diff
@@ -159,38 +175,150 @@ func TestConfig(t *testing.T) {
...
+
+		raw := testConfig(cfgFile)
+		raw["region"] = expected
+
+		c, errs := NewConfig(raw)
+		if errs != nil {
+			t.Fatalf("err: %+v", errs)
+		}
+
+		accessVal, _ := c.ConfigProvider.Region()
```
Comment (prydie) [inline builder/oracle/oci/config_test.go]: `s/accessVal/region/`

File: builder/oracle/oci/config_test.go [hunk 42]
```diff
@@ -159,38 +175,150 @@ func TestConfig(t *testing.T) {
...
+		if accessVal != expected {
+			t.Errorf("Expected ConfigProvider.Region: %s, got %s", expected, accessVal)
+		}
+	})
+
+	t.Run("ConfigProvider.KeyFingerprintOveridden", func(t *testing.T) {
```
Comment (prydie) [inline builder/oracle/oci/config_test.go]: `s/ConfigProvider.KeyFingerprintOveridden/fingerprint_overridden/`

**Pull request merged**